subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultPropertyWalker.java

```diff
@@ -53,118 +36,21 @@ public DefaultPropertyWalker(PropertyMetadataStore propertyMetadataStore) {
-                    public Object create() {
-                        try {
-                            return method.invoke(bean);
-                        } catch (InvocationTargetException e) {
-                            throw UncheckedException.throwAsUncheckedException(e.getCause());
-                        } catch (Exception e) {
-                            throw new GradleException(String.format("Could not call %s.%s() on %s", method.getDeclaringClass().getSimpleName(), method.getName(), bean), e);
-                        }
-                    }
-                });
-                return value instanceof Provider ? ((Provider<?>) value).getOrNull() : value;
-            }
-        });
-
-        public DefaultPropertyValue(String propertyName, List<Annotation> annotations, Object bean, Method method) {
-            this.propertyName = propertyName;
-            this.annotations = ImmutableList.copyOf(annotations);
-            this.bean = bean;
-            this.method = method;
-            method.setAccessible(true);
-        }
-
-        @Override
-        public String getPropertyName() {
-            return propertyName;
-        }
-
-        @Override
-        public boolean isAnnotationPresent(Class<? extends Annotation> annotationType) {
-            return getAnnotation(annotationType) != null;
-        }
-
-        @Nullable
-        @Override
-        public <A extends Annotation> A getAnnotation(Class<A> annotationType) {
-            for (Annotation annotation : annotations) {
-                if (annotationType.equals(annotation.annotationType())) {
-                    return annotationType.cast(annotation);
-                }
-            }
-            return null;
-        }
-
-        @Override
-        public boolean isOptional() {
-            return isAnnotationPresent(Optional.class);
-        }
-
-        @Nullable
-        @Override
-        public Object getValue() {
-            return valueSupplier.get();
-        }
-
-        @Nullable
-        @Override
-        public Object call() {
-            return getValue();
-        }
-
-        @Override
-        public void validate(String propertyName, boolean optional, ValidationAction valueValidator, TaskValidationContext context) {
-            Object unpacked = DeferredUtil.unpack(getValue());
-            if (unpacked == null) {
-                if (!optional) {
-                    context.recordValidationMessage(ERROR, String.format("No value has been specified for property '%s'.", propertyName));
-                }
-            } else {
-                valueValidator.validate(propertyName, unpacked, context, ERROR);
-            }
+            BeanNodeContext context = queue.remove();
+            context.visit(visitor, specFactory);
         }
     }
 
-    private class NestedBeanContext extends AbstractNestedPropertyContext<BeanNode> {
+    private class BeanNodeContext implements NodeContext<BeanNode> {
```

Comment:
This is the only implementation of `NodeContext`, I guess we can get rid of the type parameter.

Reply:
Done.